### PR TITLE
Make material, object transformation and geometry changes interactive

### PR DESCRIPTION
**Pull request by usakhelo**
Hello,
This change adds the ability to change object's material, transformation and geometry changes interactively during ActiveShade rendering mode. It also support assigning new materials and making changes during ActiveShade inside the viewport.

Thanks for review!

### CODE REVIEW

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 1]
```diff
@@ -208,17 +210,119 @@ namespace
...
                 {
-                    m_renderer->update_camera_object(m_active_camera);
+                    m_renderer->update_object_instance(node);
                     m_renderer->get_render_session()->reininitialize_render();
-                    break;
                 }
+
+                if (os.obj && os.obj->SuperClassID() == LIGHT_CLASS_ID)
+                {
+                    //TODO: update lights here
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: I suppose this is for later?

Add space after `//`.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 2]
```diff
@@ -208,17 +210,119 @@ namespace
...
+                }
+            }
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Remove redundant blank line.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 3]
```diff
@@ -208,17 +210,119 @@ namespace
...
+                m_renderer->remove_object_instance(node);
+                m_renderer->get_render_session()->reininitialize_render();
+            }
+        }
+
+        void DisplayPropertiesChanged(NodeKeyTab& nodes) override 
+        {
+            for (int i = 0, e = nodes.Count(); i < e; ++i)
+            {
+                auto test = NodeEventNamespace::GetNodeByKey(nodes[i]);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: This code doesn't do anything. Remove?

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 4]
```diff
@@ -208,17 +210,119 @@ namespace
...
+            }
+        }
+
+        void HideChanged(NodeKeyTab& nodes) override 
+        {
+            for (int i = 0, e = nodes.Count(); i < e; ++i)
+            {
+                auto test = NodeEventNamespace::GetNodeByKey(nodes[i]);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Same here.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 5]
```diff
@@ -488,6 +653,7 @@ void AppleseedInteractiveRender::EndSession()
     
     render_end(m_entities.m_objects, m_time);
 
+    // todo: fix crash when rendering in the viewport
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Is this still relevant?

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 6]
```diff
@@ -39,6 +42,110 @@
...
+
+void UpdateObjectInstanceAction::update()
+{
+    // todo: handle objects with optimized_for_instancing flag on
+    renderer::Assembly* assembly = m_session->m_project->get_scene()->assemblies().get_by_name("assembly");
+
+    if (m_session->m_object_inst_map.count(wide_to_utf8(m_node->GetName())) > 0)
+    {
+        m_session->m_object_map.erase(m_node->GetObjectRef());
+        asr::ObjectInstance* object_instance = static_cast<asr::ObjectInstance*>(m_session->m_object_inst_map[wide_to_utf8(m_node->GetName())]);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: Why is a cast necessary here?

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 7]
```diff
@@ -39,6 +42,110 @@
...
+            m_node,
+            RenderType::Default,
+            m_session->m_renderer_settings,
+            GetCOREInterface()->GetTime(),
+            m_session->m_object_map,
+            m_session->m_object_inst_map,
+            m_session->m_material_map,
+            m_session->m_assembly_map,
+            m_session->m_assembly_inst_map);
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: Remove redundant blank line.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 8]
```diff
@@ -39,6 +42,110 @@
...
+       
+        assembly->bump_version_id();
+
+        return;
+    }
+
+    if (m_session->m_assembly_inst_map.count(wide_to_utf8(m_node->GetName())) > 0)
+    {
+       m_session->m_assembly_map.erase(m_node->GetObjectRef());
+       asr::AssemblyInstance* assembly_instance = static_cast<asr::AssemblyInstance*>(m_session->m_assembly_inst_map[wide_to_utf8(m_node->GetName())]);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: Same here, why the cast?

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 9]
```diff
@@ -80,6 +86,101 @@ class CameraObjectUpdateAction
...
+  public:
+    MaterialUpdateAction(
+        renderer::Project&      project,
+        const IAppleseedMtlMap& material_map)
+      : m_material_map(material_map)
+      , m_project(project)
+    {
+    }
+
+    void update() override
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: I would suggest to implement this class (and possibly the other classes in this file) in the .cpp file instead. That should help decoupling and allow to remove some `#include` from this file, hopefully.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp [hunk 10]
```diff
@@ -44,11 +47,9 @@ namespace asr = renderer;
 
 InteractiveSession::InteractiveSession(
     IIRenderMgr*                iirender_mgr,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp]: Rename to `irender_manager` as @Mango-3 did it in the previous PR.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 11]
```diff
@@ -49,13 +50,13 @@ namespace renderer   { class Project; }
 
 class Bitmap;
 class IIRenderMgr;
+class IAppleseedMtl;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 12]
```diff
@@ -31,6 +31,7 @@
 // appleseed-max headers.
 #include "appleseedinteractive/interactiverenderercontroller.h"
 #include "appleseedrenderer/renderersettings.h"
+#include "appleseedrenderer/projectbuilder.h"
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 13]
```diff
@@ -66,15 +67,26 @@ class InteractiveSession
 
     void schedule_camera_update(
         foundation::auto_release_ptr<renderer::Camera>  camera);
+    void schedule_material_update(
+        const IAppleseedMtlMap& material_map);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: Move this arg and the one above on the same line, there's no need to wrap to a new line here.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h (hunk 11 above)
Comment (dictoon) [reply src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: Is this fwd decl really necessary?

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.h [hunk 14]
```diff
@@ -41,19 +44,42 @@
 #include "appleseed-max-common/_endmaxheaders.h"
 
 // Standard headers.
+#include <map>
 #include <vector>
 
 // Forward declarations.
 namespace renderer { class Camera; }
 namespace renderer { class ParamArray; }
+namespace renderer { class ObjectInstance; }
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp [hunk 15]
```diff
@@ -1946,3 +1837,117 @@ asf::auto_release_ptr<asr::Project> build_project(
 
     return project;
 }
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp]: Remove redundant blank line.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp (hunk 6 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: it casts from `renderer::Entity*`, I thought it's necessary. `m_object_inst_map` is `std::map<std::string, renderer::Entity*>`

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp (hunk 8 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: `m_session->m_assembly_inst_map` is also `std::map<std::string, renderer::Entity*>`
I thought I save a type by using `` for both object and assembly instances. But it's probably stupid decision.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 16]
```diff
@@ -64,8 +64,8 @@ class InteractiveSession
     void reininitialize_render();
     void end_render();
 
-    void schedule_camera_update(foundation::auto_release_ptr<renderer::Camera>  camera);
-    void schedule_material_update(const IAppleseedMtlMap& material_map);
+    void schedule_camera_update(foundation::auto_release_ptr<renderer::Camera>);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: It's unfortunate to remove the argument names, they are often extremely useful to have when reading the code.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.h [hunk 17]
```diff
@@ -75,7 +76,8 @@ struct ObjectInfo
 };
 
 typedef std::map<Object*, std::vector<ObjectInfo>> ObjectMap;
-typedef std::map<std::string, renderer::Entity*> InstanceMap;
+typedef std::map<std::string, renderer::ObjectInstance*> ObjInstanceMap;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.h]: May I suggest not to abbreviate here, for consistency's sake? I.e. `ObjectInstanceMap`

**Comment by usakhelo** [comment]
Thanks for review @dictoon ! Merging this...

**Pull request merged**